dali/core/mm/default_resources.cc

```diff
@@ -269,16 +289,8 @@ const std::shared_ptr<device_async_resource> &ShareDefaultDeviceResourceImpl(int
   if (device_id < 0) {
     CUDA_CALL(cudaGetDevice(&device_id));
   }
-  if (g_resources.device.empty()) {
-    std::lock_guard<std::mutex> lock(g_resources.mtx);
-    int ndevs = 0;
-    CUDA_CALL(cudaGetDeviceCount(&ndevs));
-    g_resources.device.resize(ndevs);
-  }
-  if (static_cast<size_t>(device_id) >= g_resources.device.size()) {
-    throw std::out_of_range(make_string(device_id, " is not a valid CUDA device index. "
-      "Shoud be 0 <= device_id < ", g_resources.device.size(), " or negative for current device."));
-  }
+  g_resources.InitDeviceResArray();
+  g_resources.CheckDeviceIndex(device_id);
```

Comment:
you don't need the mutex lock here, like in line 347?

Reply:
Not really - `Set...` is not performance-critical. Initially it was entirely under a mutex, but now I needed to move the mutex past `InitDeviceResArray` to avoid recursive locking.